src/content/filtering-string.js

```diff
@@ -0,0 +1,141 @@
+// @flow
+
+import isEqual from 'lodash.isequal';
+
+import type { IndexIntoStackTable } from '../common/types/profile';
+import type { ImplementationFilter } from './actions/types';
+
+export type FilterDescription = {|
+  implementation: ImplementationFilter,
+  substrings: string[],
+  paths: string[],
+|};
+
+export type ChargeDescription = {|
+  source: IndexIntoStackTable,
+  target: IndexIntoStackTable,
+|};
+
+export type Filter = {|
+  include: FilterDescription | null,
+  exclude: FilterDescription | null,
+  display: {|
+    invertCallstack: boolean,
+    hidePlatformDetails: boolean,
+    hide: IndexIntoStackTable[],
+    charge: ChargeDescription[],
+  |},
+  cachedString: string,
+|};
+
+function _fromFilterDescription(filter: FilterDescription): string[] {
+  const result = [];
+
+  if (filter.implementation) {
+    result.push(`implementation:${filter.implementation}`);
+  }
+
+  result.push(...filter.paths.map(path => `path:${path}`));
+  result.push(...filter.substrings.map(string => `substring:${string}`));
+
+  return result;
+}
+
+export function stringFromFilter(filter: Filter | null = null): string {
+  if (!filter) {
+    return '';
+  }
+
+  if (isEqual(filter, filterFromString(filter.cachedString))) {
+    return filter.cachedString;
+  }
+
+  const result = [];
+  if (filter.include) {
+    result.push(..._fromFilterDescription(filter.include));
+  }
+
+  if (filter.exclude) {
+    const excludedItems = _fromFilterDescription(filter.exclude).map(item => `-${item}`);
+    result.push(...excludedItems);
+  }
+
+  const display = filter.display;
+  if (display.invertCallstack) {
+    result.push('invertCallstack');
+  }
+  if (display.hidePlatformDetails) {
+    result.push('hidePlatformDetails');
+  }
+  result.push(...display.hide.map(index => `hide:${index}`));
+  result.push(...display.charge.map(({ source, target }) => `charge:${source}:${target}`));
+
+  return result.join(' ');
+}
+
+export function emptyUserFilter(): Filter {
+  return {
+    include: null,
+    exclude: null,
+    display: {
+      invertCallstack: false,
+      hidePlatformDetails: false,
+      hide: [],
+      charge: [],
+    },
+    cachedString: '',
+  };
+}
+
+const reFilterFromString = /([-+])?\b(\w+)(?::(\S+))?(?=\s|$)/g;
```

Comment:
Ahah that one is not actually complicated, it's a language, you need to find the "words" and then find what the structure is.

And like any language, you need to know how to read it first :D 

I can decompose it:

* `([-+])?`: possibly prefix with - or +
* `\b`: `\b` is the edge of a word, I like to use this a lot because this usually makes regexps faster by making them faster to fail; this is the main trick to make regexps fast. Otherwise, for example if we have a full word, a regexp engine might want to  retry the regexp starting at the letter 2, then letter 3, etc. By using \b we make it mandatory that the following `\w` need to match the first letter of the word, otherwise it doesn't match.
* `(\w+)`:  So we're fetching as much letters as possible, with at least 1 letter
* `(?::(\S+))?`: as you probably know already, `(?:` defines the start of a non-capturing group (as opposed to `(` which defines the start of a capturing group, which we have too here). `\S` is any non-space character (opposite of `\s` which is any space character). The goal here is to match the argument of a filter, like `implementation:C++` or `path:devtools/debugger`; the argument can contain anything but a space. When we'll want to match spaces we'll likely need to stop using regexps...
* `(?=\s|$)`: `(?=` defines the start of a lookahead group. This means that the regexp will match if right after the already matched string we have either a spacing character (`\s`) or the end of the string (`$`). TBH after some tests I don't think this is needed now, I don't remember why I added it so I'll remove this part ;)